design.md

```diff
@@ -0,0 +1,226 @@
+
+`-authorized-users USER-CERTIFICATE-DIR` Path to the directory containing the authorized certificates. Default: `certs/authorized`.
+
+`-listen ADDRESS:PORT` Listening address and port. Default: `localhost:9999`.
+
+## Client
+
+A successful invocation of `overseer-cli` will have a return code of zero, a non-zero value is used for error cases.
+
+### Usage
+
+`overseer-cli [-server ADDRESS:PORT] [-key PRIVATE-KEY] [-cert SERVER-CERTIFICATE] ACTION [ARGS...]`
+
+### Optional flags
+
+`-server ADDRESS:PORT` Remote server hostname or IP address and port. Default: `localhost:9999`.
+
+`-key PRIVATE-KEY` Path to the user private key in PEM format. Default: `certs/user.key`.
+
+`-cert SERVER-CERTIFICATE` Path to the remote server certificate in PEM format. Default: `certs/server.crt`.
+
+### Action flags
+
+Only one action is allowed per invocation.
+
+`-start PATH [ARGS...]` Connects to the job server at the IP/hostname given in `ADDRESS`, using the port provided in `PORT`, then starts the job at the given path (`PATH`) in the server and the arguments that follow (`ARGS`). A `JOB-ID` will be returned to uniquely identify the started job, or an error message if the execution failed.
+
+`-stop JOB-ID` Stops the job identified by `JOB-ID` and returns its exit code or an error if the provided job did not exist. It must be used to release the resources of the system.
+
+`-status JOB-ID` Returns the current state (Started or Done) of the job identified by `JOB-ID` or an error if the provided job did no exist.
+
+`-stdout JOB-ID` Writes the standard output of the given job to the standard output of this process, or returns an error if another instances of this process is already connected.
+
+`-stderr JOB-ID`	Writes the standard error of the given job to the standard output of this process, or returns an error if another instance of this process is already connected.
+
+### Example session
+
+```
+$ overseer-cli -server localhost:8888 -key my-key.key -cert server.crt -start echo Hello
+UNIQUE-JOB-ID
+
+$ overseer-cli -server localhost:8888 -key my-key.key -cert server.crt -status UNIQUE-JOB-ID
+Done
+
+$ overseer-cli -server localhost:8888 -key my-key.key -cert server.crt -stdout UNIQUE-JOB-ID
+Hello
+
+$ overseer-cli -server localhost:8888 -key my-key.key -cert server.crt -stop UNIQUE-JOB-ID
+0
+```
+
+
+## Resource control
+
+There are many ways to control the resources on a Linux system, alternatives considered were: `systemd-run`, `cpulimit`, `nice`, `ionice`,`cgroups`; but the `prlimit()` system call was chosen because it satisfies the requirements (being a system call and being able to enforce all the necessary limits). The proposed solution is to start the given job and then set the limits, the drawback being that the new process could exceed the resources before the limits are set, another drawback is that there is no reliable way to tell if a process exceeded the limits and got killed or not.
```

Comment:
I don't think rlimit allows you to limit I/O throughput.

Reply:
> the drawback being that the new process could exceed the resources before the limits are set

Also if given job will fork before setting the limit it can bypass the resource control.